lib/tty/pager/system.rb

```diff
@@ -123,26 +124,69 @@ def initialize(**options)
       #
       # @api public
       def page(text, &_callback)
-        return text unless output.tty?
+        write(text)
+        wait
+      end
+
+      # Spawn the pager process
+      #
+      # @return [PagerIO]
+      #   A wrapper for the external pager
+      #
+      # @api private
+      def spawn_pager
+        # In case there's a previous pager running:
+        wait
 
         command = pager_command
         out = self.class.run_command(command)
         # Issue running command, e.g. unsupported flag, fallback to just command
-        if !out.empty?
+        unless out.empty?
           command = pager_command.split.first
         end
 
-        pager_io = open("|#{command}", 'w')
-        pid      = pager_io.pid
+        PagerIO.new(command)
+      end
 
-        pager_io.write(text)
-        pager_io.close
+      # Send text to the pager process. Starts a new process if it hasn't been
+      # started yet. Returns false if the pager was closed.
+      #
+      # @return [Boolean]
+      #   the success status of writing to the pager process
+      #
+      # @api public
+      def write(text)
+        @pager_io ||= spawn_pager
+        @pager_io.write(text)
+      end
+      alias_method :<<, :write
 
-        _, status = Process.waitpid2(pid, Process::WNOHANG)
-        status.success?
-      rescue Errno::ECHILD
-        # on jruby 9x waiting on pid raises
-        true
+      # Send a line of text, ending in a newline, to the pager process. Starts
+      # a new process if it hasn't been started yet. Returns false if the pager
+      # was closed.
+      #
+      # @return [Boolean]
+      #   the success status of writing to the pager process
+      #
+      # @api public
+      def puts(text)
+        @pager_io ||= spawn_pager
+        @pager_io.puts(text)
+      end
+
+      # Stop the pager, wait for the process to finish. If no pager has been
+      # started, returns true.
+      #
+      # @return [Boolean]
+      #   the exit status of the child process
+      #
+      # @api public
+      def wait
```

Comment:
I will come clean here. I'm not totally sold on `wait` method. Why? This name feels like an implementation detail. Does a user need to know that they are waiting for a process?  The name breaks down a little for me in the basic pager implementation. Because of this, I'm leaning towards the `close` method. Fundamentally, this is the intention, open a pager, keep on writing and when you done close it. This also happens to mimick the `File`.

```ruby
pager = TTY::Pager.new
pager << "I know not all that may be coming"
pager << "but be it what it will, I'll go to it laughing."
pager.close
```
If user doesn't `close` the pager the expectation is that it will remain waiting. Similarly, we could allow for class-level method like `page` or `open` with a block that would close things automatically for you:

```ruby
TTY::Pager.page(command: "less -R") do |pager|
  pager << "I know not all that may be coming"
  pager << "but be it what it will, I'll go to it laughing."
end
```

Then we could allow for paging an entire file with one line:

```ruby
TTY::Pager.page(File.open("large-file.txt"), command: "less -R")
```

I think there is a significant benefit to making things similar/familiar to what Rubyists may already experience. Any thoughts?

Reply:
Yes, this makes perfect sense. Considering the need to start and wait, a block form makes a lot more sense as the "canonical" way to interact with the pager. And `close` is definitely a better name.

With that, and the write and puts aliases, I'm thinking more and more that instead of a boolean result, an exception would be much better to indicate that the pager has been closed. Particularly since it means the block form could just handle it automatically in a way that a `return false` couldn't.

So, I've implemented that, adding a `try_write` method for finer-grained control. Although I don't think it's that necessary -- I'd be happy to remove it. I could also try using `throw` instead of `raise` since it's debatable whether the pager process being closed is an "error" (the process does go away while the program is working, after all), or "control flow" (it's an expected thing that could happen because of user interaction). For all the dislike of using exceptions as control flow, they're slightly more familiar to users, so ¯\_(ツ)_/¯.

Let me know what you think. 

I've yet to write some tests for the block form. I'll get around to it. One thing I want to mention as an aside is that I feel that separating unit tests by methods per file feels a bit limiting. Multiple methods could easily be tested in one example, like `write` and `puts`, which now duplicate quite a lot of tests. But this is my own opinion, and I'm happy to leave things as they are. As you say, probably not the best idea to make too many changes in one PR.

Reply:
The current "lorem" example:

``` ruby
lorem = <<~EOF
  ... lots of text ...
EOF

TTY::Pager.page do |pager|
  10.times do
    pager.write(lorem)
    sleep 1
    pager.write "---------------------\n"
  end
end
```